pkg: lock utop dev tool in separate scheduler run

gridbugs opened the pull request:
Locking the dev tool within the same scheduler run as building utop causes the memoized filesystem functions to not be able to see utop's lockdir.

gridbugs committed:
pkg: lock utop dev tool in separate scheduler run

Locking the dev tool within the same scheduler run as building utop
causes the memoized filesystem functions to not be able to see utop's
lockdir.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

rgrinberg commented:
I'm not sure what's going here, but the description doesn't appear to make sense. Calling Scheduler.run twice does not change the state of the memoized functions.

Alizter commented:
@gridbugs Do you have a reproduction case for the issue?

gridbugs commented:
@rgrinberg ah ok my bad. The issue this fixes is that `DUNE_CONFIG__LOCK_DEV_TOOL=enabled dune utop` would create a lockdir for utop but then would be unable to see the lockdir when building utop, but running the command for a second time worked. My mental model for Fs_memo was incorrect. I'm looking again to see why this change fixed the issue.

gridbugs commented:
I've reverted this and fixed the issue without a separate scheduler run here: https://github.com/ocaml/dune/pull/12073. I assumed that `Fs_memo.dir_exists` would memoize its result and then miss the creation of the lockdir. I'm still not sure whether it does or not but it's moot because I was checking for the existence of the lockdir in a lazy memo, and I didn't realize it was being forced as a side effect of checking whether the utop exe exists.
